core/src/main/java/org/elasticsearch/http/netty/NettyHttpServerTransport.java

```diff
@@ -343,12 +344,23 @@ protected void doStart() {
             }
         }
 
-        if (0 == publishPort) {
-            throw new BindHttpException("Publish address [" + publishInetAddress + "] does not match any of the bound addresses [" + boundAddresses + "]");
+        // if no matching boundAddress found, check if there is a unique port for all bound addresses
+        if (publishPort < 0) {
+            final IntSet ports = new IntHashSet();
+            for (InetSocketTransportAddress boundAddress : boundAddresses) {
+                ports.add(boundAddress.getPort());
+            }
+            if (ports.size() == 1) {
+                publishPort = ports.iterator().next().value;
+            }
+        }
+
+        if (publishPort < 0) {
+            throw new BindHttpException(SETTING_HTTP_PUBLISH_PORT.getKey() + " is ambiguous (bound addresses: " + boundAddresses + "). " +
```

Comment:
SETTING_HTTP_PUBLISH_PORT is likely empty right? In that case I think this message will be confusing. Maybe something like "can't auto-resolve http publish port, multiple bounded ports and none of them matched the publish host. Please specify a unique port by setting" + SETTING_HTTP_PORT.getKey() + " or " + SETTING_HTTP_PUBLISH_PORT.getKey()